### ISSUE TITLE
currently no fields are being translated

### ISSUE THREAD

**Issue by Puddingboy** [AnimalDesign/bolt-translate#74]
bolt 3.1.0 and boltforms+translate fresh install(latest)

During my setup I first added
```
locales:
    nl_NL:
        label: Nederlands
        slug: nl
    en_GB:
        label: English
        slug: en
    de_AT:
        label: Deutsch
        slug: de
```

Only after updating the db I changed my default locale to `nl_NL`

currently no field is being translated and no db update is being offered. looking for suggestions to fix this.

`is_translateable: true` does not fix it. Can only reinstalling the plugin fix this?

Changes on translation tabs are not being saved nor submitted

**Comment by lswest** [comment]
Same issue here.  On some pages, I even get the error "could not save changes".  On others, it saves without complaint (but on a refresh loses all changes).  2 things I've noticed though:

1.  When I open a contenttype entry to edit it, it appears with "English" in the dropdown menu first (instead of the default language)
2. I uninstalled the extension, cleaned up the database, and then re-installed it.  Unfortunately, it didn't change anything.  My database also has rows for slug, en_slug, de_slug, and en_data/de_data.  But no such entries for the body (which is also set with `is_translateable`).

My assumption is that the changing of default languages has resulted in issues in the database which are neither caught by bolt, nor fixed by removing the plugin.

**Comment by Puddingboy** [comment]
lswest removing the plugin won't remove the db entries. the db entries need to be removed manually. 
I just checked my db and this is my current table structure and fill: http://puu.sh/r3XLz/095266166d.png
It seems that the tables are being created but not filled.

**Comment by lswest** [comment]
@Puddingboy Removing the plugin didn't, but the interim database check did seem to remove some columns at least.

I've also just run through an update, but no errors appear in my MySQL log or PHP.  I enabled the bolt-own system debugger, and (after removing the authentication events), posted the contents [here](https://gist.github.com/lswest/69627301e98ccf85d3756b8194caef6f)

From what I can gather, it seems that there is an SQL error that occurs - and my best guess is that it's caused by the hyphen (i.e. that the SQL isn't being escaped properly).  I also see these errors in red within Bolt, if I downgrade the bolt-translate extension to 3.1.1, or 3.1.0.

**Comment by calvinchoy** [comment]
I am having the same problem. Running Bolt 3.0.12 and latest version of this extension. It seems like the translation(s) are being overwritten when switching from default locale.

I first started with the default en_gb locale, and added the nl_nl locale as secondary. Everything seems to work fine. However, when I updated the configuration files and make the nl_NL locale default, all the translations are gone and "replaced" by content of the primary locale.

I also notice in the database that the locale_data columns are not being filled or only filled for one locale. the other ones is always empty.

**Comment by lswest** [comment]
Just a quick update - I've downgraded Bolt to 3.0.12 (as suggested in issue #72).  This seems to have fixed the issue.  So it seems some sort of incompatibility or error was introduced in Bolt 3.1.  Does the extension rely on the Bolt backend to write to/from the database?  Therefore a change there might be causing this.  Perhaps a question for @bobdenotter.

**Comment by calvinchoy** [comment]
Well, I've been using Bolt 3.0.12 from the start together with this plugin. I got the same problem after I switched the default locale. Now I can't seem to undo the problem by switching the default locale back. The second locale will always receive a "cannot save item" error. Only the primary locale can be saved, but then the second will be overwritten with the same content. If you look at the database columns, it is actually not saving/deleting existing content.

I will try to see if I can reproduce the problem using a fresh database.

**Comment by SahAssar** [comment]
@lswest @calvinchoy @Puddingboy Sorry for being late on this. I'm looking through it now.

I'll hopefully have a release later tonight to fix this, otherwise I'll comment here.

@lswest The question mark in SQL queries represents a variable in a prepared statement, so it's replaced with a value when run. Also the extension stores all data in one column per locale, and the slug in one column per locale. So the schema you are seeing seems correct.

**Comment by SahAssar** [comment]
Okay, this isn't exactly a quick fix... I talked with ross on slack and the way that 3.1+ works (which seems like the correct way to me although it makes my life harder currently) makes this a bit harder than I though.

@madc I think we should tag a release that is capped at 3.0.X (and therefore removing the 3.X.X releases) and then work on getting a fix for 3.1, but that will almost certainly require more user configuration than the 3.0.X release required. You okay with that?

**Comment by SahAssar** [comment]
Oh, and for future reference I'm leaving this here: https://github.com/bolt/bolt/pull/5684/commits/231cd0fc3225c994079a5e914aeb9923699f1fdd which was the commit that corrected the "hack" that I was using.

**Comment by lswest** [comment]
@SahAssar Thanks for explaining the question mark!

I'll see if working with 3.0.12 will let me finish the site I'm working on - though the slug translations don't seem to be working properly (like in issue #70).  In any case, I appreciate the effort!

**Comment by SahAssar** [comment]
Alright, so, what is now in dev-master should work, but it also requires you to manually add two hidden fields to each contenttype that you want to translate, see the readme. If someone could try it out for me and see if it works as expected for you guys that'd be great... I had to do it this way since adding the fields via `$app['config']->set()` did not seem to work...

**Comment by lswest** [comment]
I was going to test out the dev-master version, but am receiving this error: `Could not find package animal/translate at any version for your minimum-stability (dev). Check the package spelling or your minimum-stability` - this is after setting stability to dev in config.yml.  Both Bolt 3.0.12 and 3.1.1.

**Comment by SahAssar** [comment]
yeah, getting the same here... I have no idea why, but you can test it by just overwriting the old version with the new files or doing a git pull in that folder IIRC.

**Comment by lswest** [comment]
`git pull` worked to update it, but after adding the hidden fields, I now see this error:

`Es gab einen Fehler beim Updaten von `bolt_ueber_uns`. Der Error lautet An exception occurred while executing 'CREATE INDEX IDX_A82C6B3ACAAC50AC ON bolt_ueber_uns (deslug)': SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'deslug' used in key specification without a key length`

**Comment by SahAssar** [comment]
Ah, so mysql doesn't like that `index: true`... You can just remove that from the config then, it will be a bit slower, but I don't know of any way to tell bolt to create varchar columns instead of text ones...

**Comment by lswest** [comment]
Removing index: true does remove the error messages.  After that, I can save any of the content types I want, and they appear under the correct URL.  I haven't tested it on many entries, but before it didn't work at all.  This was tested in Bolt 3.1.1

Should this fix impact the slug translations too?  If so, that doesn't seem to be working as of yet.

**Comment by SahAssar** [comment]
I have tested it with slug translations too when I helped @evertalbers with the same issue in #76, so it should work... You probably need to resave the slugs for it to work though, and keep in mind that if require_slugs is on it needs saved translated slugs to work.

**Comment by lswest** [comment]
I activated the translated_slugs option in the config, and no page besides the home page (`/en` or `/de`) will load, citing the following error: `Symfony \ Component \ HttpKernel \ Exception \ HttpException
Page pages/ not found.`  This is true for both pages without translated slugs, as well as pages I've translated the slugs for.  In both languages.  I also ran another `git pull` just to make sure I was up to date.  If you need more information, I can open a new issue for this particular problem (to keep this issue thread clean-ish).

**Comment by SahAssar** [comment]
@lswest Are you on our slack or IRC? It might be easier to solve this in real time. Then I can incorporate that into the release I'm hoping to drop soon.

**Comment by lswest** [comment]
@SahAssar I'm lswest on freenode IRC

**Comment by SahAssar** [comment]
Posting the chatlog from IRC here, It seems the issue was still the move to dev-master, a git pull did not do the trick.

I'll work on releasing a translate 4.0 this week, the version jump is because it will not be compatible with our translate 3.X releases. **Just to be clear: this is a translate 4.0 release, not a bolt 4.0 release that I am talking about**. We happened to be tracking the bolt version numbers pretty closely lately, but that is just a coincidence.

Translate 3.X will only work on bolt 3.0.X, not bolt 3.1+. Translate 4.0 should work on all bolt 3.X, but translations made under translate 3.X will not be automatically moved to the new db scheme. One can do this manually by just renaming to columns to not include the underscore (and adding the hidden fields) before upgrading, but for small sites it might be easier to just copy-paste the content.

I'll close this issue when we have a release compatible with bolt 3.1+

```
19:26	SahAssar	Could you possibly pastebin or otherwise show me your contenttypes.yml and your routing.yml?
19:26	lswest	Sure, I'll drop them in a gist
19:27	lswest	https://gist.github.com/lswest/dbe03840b47fbd29689f34f48b37f556
19:28	lswest	The contenttypes file is pretty long, but most of the content types use the fields from pages
19:29	SahAssar	Seems fine... You are on 3.1.2 of bolt, right?
19:30	lswest	No, 3.1.1
19:31	SahAssar	Alright, that should work too... Sorry to ask this, but is there any way that I can get a login to bolt to check a few things?
19:31	SahAssar	If you're not fine with it I completely understand.
19:31	lswest	I don't have an issue with it, but the test version with the dev-master and bolt 3.1.1 is only a local copy
19:32	SahAssar	Ah, right. Hold on a sec.
19:33	SahAssar	Under the extend screen it's showing "dev-master (392559)", right?
19:34	lswest	Nope, the extend screen still says 3.1.2
19:35	SahAssar	Right, try the manage -> install different version -> Show all versions -> dev-master
19:35	lswest	I just ran an update to dev-master through the extend page, and it worked this time. Now it shows dev-master, and I'll test the links again briefly
19:35	SahAssar	Ah, great :)
19:36	lswest	However, after setting translate_slugs to true, the pages still aren't working
19:36	lswest	With translate_slugs off, the links load fine (albeit untranslated)
19:37	lswest	Looking at the routing page of the profiler, the correct path is selected (pages /{_locale}/{slug}) - it just fails to find the actual page
19:37	SahAssar	Try turning it on, and then going to the record in the backend and resaving it in each of the locales you have.
19:38	SahAssar	That's basically what I had to do for evertalbers.
19:38	lswest	That seems to have worked
19:38	lswest	I was only resaving in the language I was trying to test for
19:39	lswest	One last question though - I have a bunch of contenttypes (as you saw). Any way I can also translate the contenttypeslug in the URL?
19:40	SahAssar	lswest: Well, sorta, with routing.yml, but not automatically.
19:40	lswest	That's fine with me
19:41	lswest	If you had a quick example, that would be great, otherwise I'll do my best to figure it out on my own!
19:43	SahAssar	This should work, but I haven't tried it: https://gist.github.com/SahAssar/7534a9b48ec4cd4986939f988fdc5b79
19:43	lswest	Ah, that's how you work the locale in. I'll test it out right now, and let you know.
19:43	SahAssar	Basically, in your routing.yml set a default language, and turn of the automatic routing that the translate extension does.
19:44	SahAssar	All the automatic routing does is prefix every route with a /{_locale}/ param
19:46	lswest	Which is fair - but wouldn't this also impact the "use_accept_language_header"?
19:46	SahAssar	It shouldn't IIRC... But I wasn't the one implementing that, so I can't say for sure.
19:47	SahAssar	Lemme check
19:47	lswest	Sure, thanks
19:47	lswest	this would also help pretty up the URLs a bit
19:48	SahAssar	Yeah, a custom routing also allows you to use subdomains for languages like de.example.com and en.example.com
19:49	lswest	I'm not entirely sure how I'd work that into the routing, but I'll worry about that at a later date
19:49	SahAssar	Seems like the use_accept_language_header should be fine with a custom routing, but if it doesn't work please open an issue. I've had bad-to-no internet for a little while but, I should be back fixing issues now :)
19:49	lswest	Sure, no worries. You've been super fast and very efficient compared to some other projects I've opened bug reports with!
19:50	SahAssar	here is an example for the future for hostname routing: http://symfony.com/doc/current/routing/hostname_pattern.html
19:50	lswest	Thanks for the help - and the work on the extension
19:50	lswest	Wow, that was prompt. Thanks!
19:50	SahAssar	Btw, mind if I post this log in the issue as a conclusion for the issue? It might help others.
19:50	lswest	Not at all
19:51	SahAssar	Great, thanks :)
19:51	lswest	No problem!
```

**Comment by lswest** [comment]
I just tried creating a new page (called 404) in order to have a nice-looking error page.  However, when saving the page (from the create page screen), I'm greeted with this error:

```
An exception occurred while executing 'INSERT INTO bolt_pages (slug, datecreated, datechanged, datepublish, datedepublish, username, ownerid, status, templatefields, locale, title, body, dedata, deslug, endata, enslug) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["", "2016-09-21 08:31:31", "2016-09-21 08:31:31", "2016-09-21 08:31:24", null, "", "1", "published", "[]", "", "", "", "", "", "{\"title\":\"404\",\"slug\":\"page-404\",\"body\":null}", "page-404"]:\n
\n
SQLSTATE[HY000]: General error: 1364 Field 'image' doesn't have a default value
```
A few things I noticed right off the bat - there isn't an image field in my definition of Pages, and the slug field is empty ("").  No errors while editing pages that already exist.

**Comment by Puddingboy** [comment]
As far as I know the 404 page has been an issue for longer. On the version I was using till recently it threw me a bolt error page because the 404 did not exist(yup true) but after I did create a 404 page, it simply shows a white page with no content(page has no head or body)

**Comment by SahAssar** [comment]
The 404 issue should have been solved in #48 IIRC.

That error that you see @lswest looks completely unrelated though... Did you perhaps try to save in a non-default locale before saving in the default locale?

**Comment by lswest** [comment]
@SahAssar I was indeed doing that, but after changing the locale in the URL (as the dropdown doesn't appear when creating a new entry, at least not for me), the same error occurred when trying to save.

**Comment by evertalbers** [comment]
I think I have the same problem here, but no time to look at it tonight I'm afraid. If you still have the login somewhere, you're welcome.

**Comment by SahAssar** [comment]
hmm... Did you actually back out of the screen and click the "new page" button, or just change the query string? (It shouldn't matter, but who knows what might we weird when saving in non-defaults in the new system. The old one could do it, but I don't think the new one can.)

**Comment by lswest** [comment]
@SahAssar I tried both approaches, and the same error occurred each time.  However, I did notice one small change from the original error - when doing it in the default locale (de), the "id" field is filled in, while it was empty when doing it in the non-default.  I've not got time to do more investigating today, but I'll check the database tomorrow, and see if there isn't something odd going on there, and also try a different content type.

**Comment by SahAssar** [comment]
@lswest If you don't find the error could you post a separate issue with instructions to reproduce? It sounds to me like it's unrelated to this issue, especially since this issue should be solved by running dev-master as you are now doing...

**Comment by SahAssar** [comment]
@lswest @evertalbers Bug confirmed on @evertalbers'es site (didn't see his message before replying)... 

You both use MySQL, right? Cause on a SQLite site I can't repro it.

**Comment by lswest** [comment]
@SahAssar I'll open up a new issue if I can't find the error, for sure - and probably even if I do (whether I solve it or not).  And yes, I'm also on MySQL.  The weird part is that my content type doesn't define an image field.

**Comment by SahAssar** [comment]
@lswest I'm guessing you probably had an image field in it previously, but that the db column remains, right?

**Comment by evertalbers** [comment]
Yes, MySQL

**Comment by lswest** [comment]
@SahAssar Yup, I had a few columns I had to clean up.  I also didn't have any issues creating new entries for content types that did contain image types.  As such, I would call it a non-issue (or, at most, a confusing error message).

**Comment by SahAssar** [comment]
I just installed mariadb which should be 100% compatible with mysql IIRC, but I can't reproduce the issue on there...

@lswest Do you mean that the issue disappeared when you removed the old columns?

**Comment by lswest** [comment]
@SahAssar Yup, the issue was gone once I deleted two columns that no longer existed in the content type definition (image, and teaser).  If I recall correctly, they're included by default in Bolt's example pages.  I can give it a shot in mariadb tomorrow, if you want confirmation whether it occurs there or not.

**Comment by SahAssar** [comment]
@lswest That'd be awesome... I'd like to release this tomorrow (as soon as I've polished a few edges), and if I can get a real fix for this in that'd be nice.

**Comment by lswest** [comment]
@SahAssar I found some extra time today, so I've checked it now. The error doesn't occur in mariadb at all.  It's obviously patched something that avoids the error.  This is with the exact database dump that was causing issues for me earlier, and still contains the image and teaser columns.  The mariadb version is `10.1.17`, and MySQL is `5.7.15` (which should be the newest in both cases).  I'm using the official docker hub images for both, in case you want to check anything there.

**Comment by SahAssar** [comment]
🔥 MySQL 🔥

**Comment by SahAssar** [comment]
I guess I'll have to setup a MySQL install... But to me it sounds like this might be an issue in bolt core, in which case we'll need to solve it there.

**Comment by SahAssar** [comment]
@lswest Also, thank you so much for your help in this issue and others! It is greatly appreciated 👍

**Comment by lswest** [comment]
@SahAssar No problem, let me know if there's anything else I can do to help!

**Comment by peterboorsma** [comment]
@SahAssar Any news about this? I can't get the translations going with Bolt 3.1.3 with Translate dev-master. I followed the instructions in the readme carefully.
If you want me to test anything...

**Comment by SahAssar** [comment]
@peterboorsma I was just about to tag a release, but that you can't get it working with the dev-master seems like it's not really fixed even in dev-master. You don't seem to be active on slack right now, could you ping me when you are so we can check what's wrong? The main problem should have been fixed in dev-master and the other one seems to me like a MySQL problem in core (although I'm still looking at it).

**Comment by peterboorsma** [comment]
@SahAssar It could me me who's doing something wrong of course.. But, I can't get it working.

**Comment by peterboorsma** [comment]
@SahAssar I get this message:
An exception occurred while executing 'CREATE INDEX IDX_556FAC47662711BB ON bolt_nieuws (nlslug)': SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'nlslug' used in key specification without a key length

**Comment by bobdenotter** [comment]
Can confirm: Seeing the same error as Peter. Mysql 5.6.29.

**Comment by bobdenotter** [comment]
@SahAssar the `slug` column Bolt creates is a `varchar 128`. Shouldn't the `nlsug` and similiar columns also be `varchar 128` instead of `longtext`? 

Poking around a bit.. Will update in a minute.

**Comment by SahAssar** [comment]
@bobdenotter yeah, but since the fields are faked hidden fields I can't specify the type. It works without the index too.

I just pushed a change that adds a fake fieldtype `locale_data` that to make that column into a varchar.

When the config refactor lands in stable I can fix this for real (without the hidden fields and so on) IIRC.

**Comment by peterboorsma** [comment]
I updated to version dev-master(cc99e0) and removed the index lines in my contenttypes.yml. This seems to solve the problem. Thank you.

My contenttype now:

```
nieuws:
    name: Nieuws
    singular_name: Nieuwsitem
    fields:
        locale:
            type: locale
            group: Content
        title:
            type: text
            class: large
            group: Content
            is_translateable: true
        nldata:
            type: hidden
        nlslug:
            type: hidden
        endata:
            type: hidden
        enslug:
            type: hidden
```

**Comment by peterboorsma** [comment]
@SahAssar If I add the index: true option, Bolt keeps giving the following message. That's why I removed it.
 An exception occurred while executing 'CREATE INDEX IDX_556FAC47662711BB ON bolt_nieuws (nlslug)': SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'nlslug' used in key specification without a key length


```
nieuws:
    name: Nieuws
    singular_name: Nieuwsitem
    fields:
        locale:
            type: locale
            group: Content
        title:
            type: text
            class: large
            group: Content
            is_translateable: true
        nldata:
            type: hidden
        nlslug:
            type: hidden
            index: true
        endata:
            type: hidden
        enslug:
            type: hidden
            index: true
```

**Comment by SahAssar** [comment]
@peterboorsma See the example in the readme, the slug fields use the type `type: locale_data` to make them into the correct column type, just change your config to that and it should work.

```
[...]
dedata:
    type: hidden
deslug:
    type: locale_data
    index: true
endata:
    type: hidden
enslug:
    type: locale_data
    index: true
[...]
```

**Comment by peterboorsma** [comment]
My bad. You are right. Changed it, works fine now.

```
nieuws:
    name: Nieuws
    singular_name: Nieuwsitem
    fields:
        locale:
            type: locale
            group: Content
        title:
            type: text
            class: large
            group: Content
            is_translateable: true
        nldata:
            type: hidden
        nlslug:
            type: locale_data
            index: true
        endata:
            type: hidden
        enslug:
            type: locale_data
            index: true
```
